src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperUtil.java

```diff
@@ -237,37 +223,184 @@ public static Object deserializeStoredVector(BytesRef storedVector, VectorDataTy
      * @return expected vector length
      */
     public static int getExpectedVectorLength(final KNNVectorFieldType knnVectorFieldType) {
-        int expectedDimensions = knnVectorFieldType.getDimension();
-        if (isModelBasedIndex(expectedDimensions)) {
+        int expectedDimensions;
+        if (ANNConfigType.FROM_MODEL == knnVectorFieldType.getAnnConfig().getAnnConfigType()) {
             ModelMetadata modelMetadata = getModelMetadataForField(knnVectorFieldType);
             expectedDimensions = modelMetadata.getDimension();
+        } else {
+            expectedDimensions = knnVectorFieldType.getAnnConfig()
+                .getDimension()
+                .orElseThrow(() -> new IllegalArgumentException("Dimension not found"));
         }
         return VectorDataType.BINARY == knnVectorFieldType.getVectorDataType() ? expectedDimensions / 8 : expectedDimensions;
     }
 
-    private static boolean isModelBasedIndex(int expectedDimensions) {
-        return expectedDimensions == -1;
-    }
-
     /**
      * Returns the model metadata for a specified knn vector field
      *
      * @param knnVectorField knn vector field
      * @return the model metadata from knnVectorField
      */
     private static ModelMetadata getModelMetadataForField(final KNNVectorFieldType knnVectorField) {
-        String modelId = knnVectorField.getModelId();
+        ModelMetadata[] modelMetadata = new ModelMetadata[1];
```

Comment:
This looks as a strange pattern to me :)
Why not
```
String modelId = knnVectorFieldgetAnnConfig().getModelId().getOrThrow(...);
ModelMetadata modelMetadata = modelDao.getMetadata(modelId);
if (!ModelUtil.isModelCreated(modelMetadata) {
  throw ...
}
return modelMetadata;
```

Reply:
Yeah agree. Refactoring